[[ ScriptExtensions ]] Make script extensions work like lcb extensions

livecodeali opened the pull request:
To provide a bit of uniformity in the way script-library extensions
are built, packaged, and included in the IDE and standalones, this
patch rejigs some gyp files and the installer manifest to build the
script libraries into the packaged_extensions folder and include them
in the IDE Extensions folder in the installed LiveCode.

Additionally this patch:
- Tweaks the docs parser to accept title, version and author metadata
  at the library level. This info is used as the extension metadata
  in lieu of a manifest
- Adds the library level docs as metadata for the current set of
  script libraries
- Adds a script library extension packaging script, which builds
  optionally protected script libraries into the packaged_extensions
  folder

livecodeali commented:
WIP until it resolves the 'requires' metadata issue. We should be able to use essentially the same mechanism as lcb extensions - specify the requirements as a list `name (type), ...`, topologically sort to determine load order.

montegoulding commented on util/package_script_extension.livecodescript:
```diff
@@ -0,0 +1,79 @@
+command packageScriptExtension pSourceFolder, pSourceFileName, pTargetFolder, pRemoveSource   
+
+   local tTargetDir, tBuildTarget, tZip
+   put "com.livecode.script-library." & pSourceFileName into tTargetDir
```
can you use `the short name of stack` here?

livecodeali replied:
yes!

livecodeali replied:
Hmm, actually probably not yet, as the gyp has the dir hard-coded for output

montegoulding commented on util/package_script_extension.livecodescript:
```diff
@@ -0,0 +1,79 @@
+   
+   if pRemoveSource is "true" then
+      put return & "rm" && quote & tBuildTarget & slash & \ 
+         pSourceFileName & ".livecodescript" & quote after tShell
```
I'm not sure if this makes any sense unless there's a compiled version created here (binary/password protected?)

montegoulding replied:
Scratch that I see there is!

montegoulding commented on Installer/package.txt:
```diff
@@ -450,6 +450,23 @@ component Extensions
 	into [[ToolsFolder]]/Extensions place
 		rfolder macosx:packaged_extensions/com.livecode.library.scriptitems
 
+    into [[ToolsFolder]]/Extensions place
```
I think the indenting is wrong here. These files are _very_ fussy about indenting

livecodeali committed:
[[ ScriptExtensions ]] Add metadata to script extensions

livecodeali committed:
[[ ScriptExtensions ]] Build LCS extensions in the same way as LCB extensions

livecodeali committed:
[[ ScriptExtensions ]] Include packaged script extensions in installer

livecodeali committed:
[[ ScriptExtensions ]] Support script extensions metadata in library-level docs

livecodeali committed:
[[ ScriptExtensions ]] Move icon calculation to ide extension library

livecodeali committed:
[[ ScriptExtensions ]] Load included script libraries on standalone startup

livecodeali committed:
[[ ScriptExtensions ]] Change variable name when checking stacks for extensions

livecodeali commented:
Depends on https://github.com/livecode/livecode-ide/pull/1773

livecodeali committed:
[[ ScriptExtensions ]] Rename script library extensions

livecodeali committed:
[[ Docs ]] Remove now-reserved param names from docs

livecodeali committed:
[[ Extensions ]] Use packaging script for lcb extensions
In particular, generate lcb extension docs in extension package.

livecodeali committed:
[[ Tests ]] Test extension docs

livecodeali committed:
[[ Extensions ]] Fix extension docs to pass tests

livecodeali committed:
[[ DocsParser ]] Don't treat /**/ as doc block

livecodeali committed:
[[ DocsParser ]] Add 'module' type for lcb modules not intended to have lcs api

livecodeali committed:
[[ DocsParser ]] Use similar types as placeholders

livecodeali committed:
[[ DocsParser ]] Transfer post-parse library elements to library api entry

livecodeali committed:
[[ Extensions ]] Ignore api.lcdoc files
These are autogenerated when the extensions are packaged, or when
the IDE is run from the repo. Non-inline extension docs can still
be added using git add -f, or using a name other than api.lcdoc.

livecodeali committed:
[[ Extensions ]] Fix extensions build dependencies

PaulMcClernan committed:
Update URL-scheme.lcdoc

Include not Unclude, LOL. There is also issue on LC9 dp9 where binfile is not displayed at the end of the description for some reason but I don't see where that might be coming from here. Maybe bindle is not setup as a valid keyword?

bwmilby committed:
[[Bug 20515]] Segmented control widget highlights off

`hintLower` handler should not subtract 0.5 from pBound.  This was resulting in an un-highlighted line one pixel wide at the right/bottom side of a cell depending on how the rounding went.

PaulMcClernan committed:
Update URL-scheme.lcdoc

binfile & resfile changed from glossary to keywords references

bwmilby committed:
[[Bug 20515]] Add note

Prevent segmented control fill from ending one pixel to the left/bottom of segment.

montegoulding committed:
[[ Cleanup ]] Add missing field initializer

livecodeali committed:
[[ ScriptExtensions ]] Change revLoadLibrary and revUnloadLibrary
to extensionInitialize and extensionFinalize respectively. Also
change revLibraryStartupScript to extensionStartupScript

livecodeali committed:
[[ Docs ]] Fix a couple of issues
- Some docs go from index 0 upwards, so iterate correctly
- log an error when no name is found, rather than throwing

livecodeali committed:
[[ ScriptExtensions ]] Add extension packager script

livecodeali committed:
[[ ScriptExtensions ]] Update call to build extensions script

montegoulding committed:
[[ LCIDLC ]] Fixes for building on windows

livecodeali committed:
[[ ScriptExtensions ]] Move dependent settings to target

livecodepanos committed:
Add MacOS High Sierra and Android Oreo in the supported platforms

livecodeali committed:
[[ ScriptExtensions ]] Fix packaging script
Remove pushd as it is a bash extension and therefore does not
always work when used with a block of shell script executed from LC

livecodeali committed:
[[ ScriptExtensions ]] Change some remaining revLoadLibrary occurrences

livecodeali committed:
[[ ScriptExtensions ]] Change some extension stack names

livecodeali committed:
[[ ScriptExtensions ]] Ensure test stacks pass revLoadLibrary when appropriate

livecodeali committed:
Update version

livecodeali commented:
Here is the Travis test branch which now passes:
https://github.com/livecode/livecode/pull/6075

This build will hopefully contain all the associated changes, so needs to be checked before these PRs are merged https://vulcan.livecode.com/builders/release-commercial/builds/1665

livecodeali commented:
IDE PR: https://github.com/livecode/livecode-ide/pull/1773

montegoulding commented:
@livecodeali just incase you have't seen it there's conflicts to resolve here

livecodeali commented:
Gah, again! I just resolved some

montegoulding commented:
It's pretty annoying that github doesn't send the author an email about conflicts...

livecodeali committed:
Merge branch 'develop' of github.com:livecode/livecode into script-extensions

livecodeali committed:
[[ Docs ]] Don't add docs with empty name

montegoulding commented:
@livecode-vulcan review ok 006944d

livecode-vulcan commented:
💙 review by @montegoulding ok 006944d80e891ad5a1818ef9bd9c04c0dbae1b3c

livecode-vulcan commented:
😎 test success 006944d80e891ad5a1818ef9bd9c04c0dbae1b3c

* try-community-armv6-android-api8: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api8/builds/969)
* try-community-armv6-android-api9: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-api9/builds/985)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/959)
* try-community-universal-ios-iphoneos11.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos11.1/builds/32)
* try-community-universal-ios-iphonesimulator11.1: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator11.1/builds/32)
* try-community-universal-mac-macosx10.6: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.6/builds/963)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/986)
* try-community-x86-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian7/builds/969)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/985)
* try-community-x86_64-linux-debian7: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian7/builds/968)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/978)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/1000)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/964)
